### DISCUSSION TITLE
Returning `error` in Option type ?

### DISCUSSION THREAD

**Discussion by impopular-guy** [vlang/v#17423] [General]
I can do the following on an Option type
```v
mut a := ?int(5)
a = 1
a = none
```
I can also do this
```v
fn func() ?int {
	return error('Some error')
}
a = func()
```

However, I get __`builder error`__ if I try to directly assign the error 
```v
a = error('Some error')
```

#### Questions:
The docs say that Option types may represent `none` and Result may represent error. So I was under the impression that error 
cannot be passed when we have Option return type, since Option/Result was separated a while ago. I was able to put error in the option type. 
Also the C code tells that `none` is just an `IError`. This might be useful when we want to return none as well as error from same 
function, but here we will have to manually check and propagate errors.
- Is this expected behaviour?
- Is `a = error('Some error')` a bug? Of course it is because its a builder error. But should such be allowed?

**Comment by JalonSolov** [comment]
It's expected for now, since V used to only support a single type that could return either `none` or an error.  This is deprecated, and due to be removed this year, but is still active for now to avoid breaking code.

Directly assigning an error to a variable is an error, not a bug.  Errors don't have types that can be assigned, they can only be checked in `or` blocks.

**Reply by impopular-guy** [reply]
Understood.

What to do in this situation: I have a function A that returns Option type but inside it it calls another func B that returns result type and I wish to propagate the error.

Out of curiosity i tried:
```v
fn func() !?int {  // also ?!int
	return 0
}
```
This was showing some weird errors. https://play.vlang.io/p/007db98ae4

**Reply by JalonSolov** [reply]
`!?`/`?!` is invalid syntax.  At least after the deprecation, you will be able to return an Option or a Result, but not both.

**Comment by medvednikov** [comment]
It should result in a warning about the split, will fix.
